### PR TITLE
Activity report

### PR DESCRIPTION
**Pull request by guenthermi**
This Branch add functions to let the client print a report file which include the names / locations of the generated files (with statistical information about them) after all actions got finished.

### CODE REVIEW

**Comment by mkroetzsch** [comment]
Oops, this request cannot be merged automatically right now.

File: wdtk-client/src/test/java/org/wikidata/wdtk/client/ClientTest.java [hunk 1]
```diff
@@ -83,6 +83,23 @@ public void testQuietLoggingConfig() throws ParseException, IOException {
 	}
 
 	@Test
+	public void testJsonOutput() {
+		DumpProcessingAction action = new JsonSerializationAction();
+		action.open();
+		action.close();
+		assertEquals(action.getReport(),
+				"Finished serialization of 0 EntityDocuments in file {PROJECT}-{DATE}.json");
```
Comment (mkroetzsch) [inline wdtk-client/src/test/java/org/wikidata/wdtk/client/ClientTest.java]: Why does the report contain a file-name pattern instead of a real file name?

File: wdtk-client/src/main/java/org/wikidata/wdtk/client/RdfSerializationAction.java [hunk 2]
```diff
@@ -251,4 +254,21 @@ private void printHelp() {
 					+ TASK_HELP.get(supportedTask), 75, "\n   ", true));
 		}
 	}
+
+	@Override
+	public String getReport() {
+		String message = "Finished serialization of "
+				+ this.serializer.getTripleCount() + " RDF triples in file "
+				+ this.outputDestination;
```
Comment (mkroetzsch) [inline wdtk-client/src/main/java/org/wikidata/wdtk/client/RdfSerializationAction.java]: Related to my comment on the JSON test: this should not use this.outputDestination but the same file name as is really used. This is computed in the method createRdfSerializer above. This computation should be moved to another function so that it can be fetched from several places (and this function should call insertDumpInformation to replace the placeholders).

File: wdtk-client/src/main/java/org/wikidata/wdtk/client/JsonSerializationAction.java [hunk 3]
```diff
@@ -97,4 +103,21 @@ public void close() {
...
+		String message = "Finished serialization of "
+				+ this.serializer.getEntityDocumentCount()
+				+ " EntityDocuments in file " + this.outputDestination;
+		return message;
+	}
+
+	@Override
+	public String getActionName() {
+		if (this.name != null) {
+			return name;
```
Comment (mkroetzsch) [inline wdtk-client/src/main/java/org/wikidata/wdtk/client/JsonSerializationAction.java]: should be: this.name

File: wdtk-client/src/main/java/org/wikidata/wdtk/client/DumpProcessingAction.java [hunk 4]
```diff
@@ -101,4 +101,26 @@
...
+
+	/**
+	 * Sets the actions name.
+	 * 
+	 * @param name
+	 */
+	void setActionName(String name);
+
+	/**
+	 * Returns a name which terms the action.
```
Comment (mkroetzsch) [inline wdtk-client/src/main/java/org/wikidata/wdtk/client/DumpProcessingAction.java]: Analogous to the setActionName documentation, this should read "Returns the name of the action.".

File: wdtk-client/src/main/java/org/wikidata/wdtk/client/DumpProcessingAction.java [hunk 5]
```diff
@@ -101,4 +101,26 @@
...
+	/**
+	 * Returns a report message containing information about the files which
+	 * were generated.
+	 * 
+	 * @return report message
+	 */
+	String getReport();
+
+	/**
+	 * Sets the actions name.
```
Comment (mkroetzsch) [inline wdtk-client/src/main/java/org/wikidata/wdtk/client/DumpProcessingAction.java]: Would need to be "the action's name" or "the name of the action". I would also add "If this is not set, a default name will be used."

File: wdtk-client/src/main/java/org/wikidata/wdtk/client/DumpProcessingAction.java [hunk 6]
```diff
@@ -29,16 +29,16 @@
...
- *
+ * 
  */
 public interface DumpProcessingAction extends EntityDocumentDumpProcessor {
 
 	/**
 	 * Returns true if processing requires a {@link Sites} object to be set.
 	 * This is the case for some operations that process site links.
-	 *
+	 *
```
Comment (mkroetzsch) [inline wdtk-client/src/main/java/org/wikidata/wdtk/client/DumpProcessingAction.java]: Any idea why these changes are happening? I am also seeing this on my machine and it creates a lot of pointless changes. Is Maven doing this or Eclipse?

File: wdtk-client/src/main/java/org/wikidata/wdtk/client/RdfSerializationAction.java [hunk 7]
```diff
@@ -251,4 +254,21 @@ private void printHelp() {
...
+		return message;
+	}
+
+	@Override
+	public String getActionName() {
```
Comment (mkroetzsch) [inline wdtk-client/src/main/java/org/wikidata/wdtk/client/RdfSerializationAction.java]: Could we move this to the abstract super class and use the local name of this.class() as a default? Alternatively, could we just require a default action name as a constructor parameter to the abstract super class. Then the superclass can implement getActionName() (once and for all) without having to repeat this code in each subclass. But using the local class name seems even simpler.

File: wdtk-client/src/main/java/org/wikidata/wdtk/client/Client.java [hunk 8]
```diff
@@ -151,6 +155,12 @@ public void performActions() {
 		openActions();
 		this.dumpProcessingController.processDump(dumpFile);
 		closeActions();
+		try {
+			createReport();
+		} catch (IOException e) {
+			logger.error("Could not print report file!");
```
Comment (mkroetzsch) [inline wdtk-client/src/main/java/org/wikidata/wdtk/client/Client.java]: Please add the exception message to the logged string.

File: wdtk-client/src/main/java/org/wikidata/wdtk/client/Client.java [hunk 9]
```diff
@@ -204,6 +214,32 @@ private void openActions() {
 	}
 
 	/**
+	 * Creates a report file including the results which came from the
+	 * {@link DumpProcessingAction#getReport()} methods or printing it to
+	 * stdout.
+	 * 
+	 * @throws IOException
+	 */
+	void createReport() throws IOException {
```
Comment (mkroetzsch) [inline wdtk-client/src/main/java/org/wikidata/wdtk/client/Client.java]: Should be called "writeReport". Functions that start with "create..." should return a (newly created) object.

File: wdtk-client/src/main/java/org/wikidata/wdtk/client/Client.java [hunk 10]
```diff
@@ -204,6 +214,32 @@ private void openActions() {
 	}
 
 	/**
+	 * Creates a report file including the results which came from the
+	 * {@link DumpProcessingAction#getReport()} methods or printing it to
+	 * stdout.
```
Comment (mkroetzsch) [inline wdtk-client/src/main/java/org/wikidata/wdtk/client/Client.java]: Please review this sentence (word order is mixed up). Also, it seems that you are never printing to stdout; you use the logger.

File: wdtk-client/src/main/java/org/wikidata/wdtk/client/ClientConfiguration.java [hunk 11]
```diff
@@ -252,18 +267,22 @@ public boolean getOfflineMode() {
...
 	 * to stdout.
-	 *
+	 * 
 	 * @return true if the application should not log messages to stdout
 	 */
 	public boolean isQuiet() {
 		return this.quiet;
 	}
 
+	public String getReportFilename() {
```
Comment (mkroetzsch) [inline wdtk-client/src/main/java/org/wikidata/wdtk/client/ClientConfiguration.java]: Documentation missing.

File: wdtk-client/src/main/java/org/wikidata/wdtk/client/ClientConfiguration.java [hunk 12]
```diff
@@ -95,6 +95,10 @@
 	 * Short command-line alternative to {@link #OPTION_OUTPUT_STDOUT}.
 	 */
 	public static final String CMD_OPTION_OUTPUT_STDOUT = "s";
+	/**
+	 * Short command-line alternative to {@link #OPTION_OUTPUT_REPORT}.
```
Comment (mkroetzsch) [inline wdtk-client/src/main/java/org/wikidata/wdtk/client/ClientConfiguration.java]: Wrong documentation

File: wdtk-client/src/main/java/org/wikidata/wdtk/client/ClientConfiguration.java [hunk 13]
```diff
@@ -142,6 +146,11 @@
 	 */
 	public static final String OPTION_FILTER_PROPERTIES = "fProp";
 	/**
+	 * Name of the option to print a report about the files produced by
+	 * DumpProcessingOutputActions
```
Comment (mkroetzsch) [inline wdtk-client/src/main/java/org/wikidata/wdtk/client/ClientConfiguration.java]: Misleading documentation. The option is not "to print a report" but to define the output file to print the report to. The report will only be "printed" if the option is not set, right?

File: wdtk-client/src/main/java/org/wikidata/wdtk/client/Client.java [hunk 14]
```diff
@@ -204,6 +214,32 @@ private void openActions() {
...
+	 */
+	void createReport() throws IOException {
+		StringBuilder builder = new StringBuilder();
+		for (DumpProcessingAction action : this.clientConfiguration
+				.getActions()) {
+			builder.append(action.getActionName() + ": ");
+			builder.append(action.getReport());
+			builder.append(System.getProperty("line.separator"));
+		}
+		if (this.clientConfiguration.reportFilename != null) {
```
Comment (mkroetzsch) [inline wdtk-client/src/main/java/org/wikidata/wdtk/client/Client.java]: Use the getter; don't read package private fields directly.

File: wdtk-client/src/main/java/org/wikidata/wdtk/client/Client.java [hunk 15]
```diff
@@ -204,6 +214,32 @@ private void openActions() {
...
+			File reportFile = new File(this.clientConfiguration.reportFilename);
```
Comment (mkroetzsch) [inline wdtk-client/src/main/java/org/wikidata/wdtk/client/Client.java]: use the getter

File: wdtk-client/src/main/java/org/wikidata/wdtk/client/Client.java [hunk 16]
```diff
@@ -204,6 +214,32 @@ private void openActions() {
...
+			BufferedWriter writer = new BufferedWriter(new FileWriter(
+					reportFile));
+			writer.write(builder.toString());
+			writer.close();
+		} else {
+			logger.info((builder.toString()));
```
Comment (mkroetzsch) [inline wdtk-client/src/main/java/org/wikidata/wdtk/client/Client.java]: Why are there double parentheses?

Comment (mkroetzsch) [inline wdtk-client/src/main/java/org/wikidata/wdtk/client/Client.java]: Wouldn't it be nicer to have one info log message per action instead of having each of them in one report?

File: wdtk-client/src/main/java/org/wikidata/wdtk/client/DumpProcessingOutputAction.java [hunk 17]
```diff
@@ -93,9 +93,11 @@
 	 */
 	protected Set<Closeable> outputStreams = new HashSet<>();
 
+	protected String name;
```
Comment (mkroetzsch) [inline wdtk-client/src/main/java/org/wikidata/wdtk/client/DumpProcessingOutputAction.java]: Missing documentation.

File: wdtk-client/src/main/java/org/wikidata/wdtk/client/RdfSerializationAction.java [hunk 18]
```diff
@@ -251,4 +254,21 @@ private void printHelp() {
...
+		if (this.name != null) {
+			return name;
```
Comment (mkroetzsch) [inline wdtk-client/src/main/java/org/wikidata/wdtk/client/RdfSerializationAction.java]: this.name

File: wdtk-client/src/test/java/org/wikidata/wdtk/client/ClientTest.java [hunk 19]
```diff
@@ -83,6 +83,23 @@ public void testQuietLoggingConfig() throws ParseException, IOException {
...
+	}
+
+	@Test
+	public void testRdfOutput() {
+		DumpProcessingAction action = new RdfSerializationAction();
+		action.open();
+		action.close();
+		assertEquals(action.getReport(),
+				"Finished serialization of 24 RDF triples in file null");
```
Comment (mkroetzsch) [inline wdtk-client/src/test/java/org/wikidata/wdtk/client/ClientTest.java]: This is not a robust test. The number 24 will change as soon as we change the vocabulary.

File: wdtk-client/src/test/java/org/wikidata/wdtk/client/ClientConfigurationTest.java [hunk 20]
```diff
@@ -85,6 +85,7 @@ public void testDefaultArguments() {
 		assertEquals(config.getFilterLanguages(), null);
 		assertEquals(config.getFilterSiteKeys(), null);
 		assertEquals(config.getFilterProperties(), null);
+		assertEquals(config.getReportFilename(), null);
```
Comment (mkroetzsch) [inline wdtk-client/src/test/java/org/wikidata/wdtk/client/ClientConfigurationTest.java]: General comment: assertEquals should always have the expected value in the first argument, and the tested value in the second. This is wrong in much of our existing code, but we should try to fix it when we touch test methods or create new ones.

File: wdtk-client/src/test/java/org/wikidata/wdtk/client/JsonSerializationActionTest.java [hunk 21]
```diff
@@ -60,6 +60,7 @@ public void testDefaults() {
 		assertTrue(action instanceof JsonSerializationAction);
 		assertFalse(action.needsSites());
 		assertTrue(action.isReady());
+		assertEquals(action.getActionName(), "JsonSerializationAction");
 	}
```
Comment (mkroetzsch) [inline wdtk-client/src/test/java/org/wikidata/wdtk/client/JsonSerializationActionTest.java]: Currently we are missing tests that test the actual report output. You are testing the report string, but not its writing to file. This could be added below by setting a report file for one of the tests, and making sure that it contains a line starting with the action name. It would also be useful to set a non-standard action name to test this in action as well. If you do this in only one of the methods that test JSON output, both report file and logger report will be tested.

File: wdtk-client/src/test/java/org/wikidata/wdtk/client/RdfSerializationActionTest.java [hunk 22]
```diff
@@ -38,6 +38,7 @@ public void testDefaults() {
 		assertTrue(action instanceof RdfSerializationAction);
 		assertTrue(action.needsSites());
 		assertTrue(action.isReady());
+		assertEquals(action.getActionName(), "RdfSerializationAction");
 	}
```
Comment (mkroetzsch) [inline wdtk-client/src/test/java/org/wikidata/wdtk/client/RdfSerializationActionTest.java]: Same comment as for JSON, but probably it's enough to test report file writing once since the code is shared.

**Comment by mkroetzsch** [comment]
@guenthermi I have added a number of inline comments. Once these are fixed, the branch should be ready to merge.

File: wdtk-client/src/main/java/org/wikidata/wdtk/client/Client.java (hunk 8 above)
Comment (guenthermi) [inline wdtk-client/src/main/java/org/wikidata/wdtk/client/Client.java]: In the prepareSites method the exception message is also missing. Should I change this there too?

File: wdtk-client/src/test/java/org/wikidata/wdtk/client/JsonSerializationActionTest.java (hunk 21 above)
Comment (guenthermi) [inline wdtk-client/src/test/java/org/wikidata/wdtk/client/JsonSerializationActionTest.java]: I don't understand what you mean with "setting a report file". The report file is created by the createReport method in the Client. The report messages consisting of the action name and the getReport result are generated in this method, too. So I don't know to test it. It is also not possible to mock this, or something like this...

File: wdtk-client/src/main/java/org/wikidata/wdtk/client/RdfSerializationAction.java (hunk 7 above)
Comment (guenthermi) [inline wdtk-client/src/main/java/org/wikidata/wdtk/client/RdfSerializationAction.java]: We could also use a Getter for an default action name. So we can move the method to the abstract super class and the default action name need not depend on the class name.

File: wdtk-client/src/test/java/org/wikidata/wdtk/client/JsonSerializationActionTest.java (hunk 21 above)
Comment (mkroetzsch) [inline wdtk-client/src/test/java/org/wikidata/wdtk/client/JsonSerializationActionTest.java]: I see, so it's another file to test. But why can't you test the method in the Client? If your code is really untestable, then it should be changed to be testable (at least I don't see any function here that cannot be tested in principle). Indeed, with the code in the client, you could even create a mock action that creates a predefined report so that you can test the exact string. But I agree that this test should be part of the client test file, not of this action's test file.

File: wdtk-client/src/main/java/org/wikidata/wdtk/client/Client.java (hunk 8 above)
Comment (mkroetzsch) [inline wdtk-client/src/main/java/org/wikidata/wdtk/client/Client.java]: Yes, please do.

File: wdtk-client/src/main/java/org/wikidata/wdtk/client/RdfSerializationAction.java (hunk 7 above)
Comment (mkroetzsch) [inline wdtk-client/src/main/java/org/wikidata/wdtk/client/RdfSerializationAction.java]: Yes, that would be an acceptable solution too. I think, however, that such additional get methods are not necessary here: I don't see why one should return something else than the class name. It seems like a good default, and subclasses could still overwrite it by calling the setter in their constructor with another string.

File: wdtk-client/src/main/java/org/wikidata/wdtk/client/DumpProcessingAction.java (hunk 6 above)
Comment (guenthermi) [inline wdtk-client/src/main/java/org/wikidata/wdtk/client/DumpProcessingAction.java]: I think this is caused by the eclipse auto-formatter. Maven do not make such changes.

Comment (mkroetzsch) [inline wdtk-client/src/main/java/org/wikidata/wdtk/client/DumpProcessingAction.java]: Could you try to switch this off in your Eclipse? When I safe a file, it removes these spaces. The formatter (SHIFT+CTRL+F) does not change the spaces in my installation. As it is now, we are getting a lot of distracting changes in the diffs based on this.

File: wdtk-client/src/main/java/org/wikidata/wdtk/client/Client.java [hunk 23]
```diff
@@ -204,6 +215,35 @@ private void openActions() {
...
+				.getActions()) {
+			if (this.clientConfiguration.getReportFilename() != null) {
+				builder.append(action.getActionName() + ": ");
+				builder.append(action.getReport());
+				builder.append(System.getProperty("line.separator"));
+			} else {
+				logger.info(action.getActionName() + ": " + action.getReport());
+			}
+		}
+		if (this.clientConfiguration.getReportFilename() != null) {
```
Comment (mkroetzsch) [inline wdtk-client/src/main/java/org/wikidata/wdtk/client/Client.java]: Move up to the if we already have.

File: wdtk-client/src/main/java/org/wikidata/wdtk/client/Client.java [hunk 24]
```diff
@@ -204,6 +220,45 @@ private void openActions() {
...
+				outputDirectory = Paths.get(".");
+			}
+			DirectoryManager dm = DirectoryManagerFactory
+					.createDirectoryManager(outputDirectory);
+			OutputStream out = dm.getOutputStreamForFile(Paths
+					.get(this.clientConfiguration.getReportFilename())
+					.getFileName().toString());
+
+			OutputStream bufferedFileOutputStream = new BufferedOutputStream(
+					out, 1024 * 1024 * 5);
```
Comment (mkroetzsch) [inline wdtk-client/src/main/java/org/wikidata/wdtk/client/Client.java]: That's a big buffer for the three lines of report you normally have. Maybe just use the default.
